sdk/python/kfp/dsl/_metadata.py

```diff
@@ -199,6 +205,10 @@ def _extract_component_metadata(func):
 def _extract_pipeline_metadata(func):
   '''Creates pipeline metadata structure instance based on the function signature.'''
 
+  # Importing here to prevent circular import failures
+  #TODO: Change _pipeline_param to stop importing _metadata
```

Comment:
Why importing twice in this module?

Reply:
As the comment says, importing at module-level makes dependencies circular which causes import to fail. (_pipeline_param imports _metadata which imports _pipeline_param). You can try it.

In future we should probably make _pipeline_param not depend on _metadata.

Reply:
How about moving the extract_metadata functions to pipeline and component modules?

Reply:
Created an issue : https://github.com/kubeflow/pipelines/issues/1418